### DISCUSSION TITLE
Lsoda solver error, converge failing repeatedly

### DISCUSSION THREAD

**Discussion by afee36912** [nlmixr2/nlmixr2#227] [Q&A] [answered]
I repeatedly keep on getting an unhandled error message for the lsoda solver in nlmixr2. I am trying to fit an additive error PBPK model using saem, and I am not sure what is wrong. I included a screenshot below for more context, and a block of equations and conditions from the model block I am using. This model does run in RxODE2 simulations. I have used the NONMEM data I am loading in for other PK models, and they have run. How do I elect a different solver from lsoda? Is it possible that my parameter estimates are so far off that the model does not converge? Thanks again. 

I am running this in R version 4.4.0 in RStudio.
![error_message](https://github.com/nlmixr2/nlmixr2/assets/87444872/60af03ba-d18f-4232-abd3-312c62c63ce2)

# Inside the "model" block for nlmixr2 
    Rtt(0) = ksyn_tt /kdeg; # initial condition of target (none are bound initially, tight)
    Rlk(0) = ksyn_lk /kdeg; # initial condition of target (none are bound initially, leaky)
    
    # drug concentration in central compartment:
    # C is the concentration of drug in the central compartment, in nM
    C = Ac / Vp # nM cmpt2
    
    # drug concentration in tight tissue compartment:
    # Tt is the concentration of drug in the tight tissue compartment, in nM
    Tt = Att / Vtight # nM cmpt3
    
    # drug concentration in leaky tissue compartment:
    # Lk is the concentration of drug in the leaky tissue compartment, in nM
    Lk = Alk / Vleaky # nM cmpt4
    
    # drug concentration in lymph tissue compartment:
    # Lp is the concentration of drug in the lymph tissue compartment, in nM
    Lp = Alym / Vlymph # nM cmpt5
    
    # ODEs
    d/dt(Ad) = -ka * Ad # # Ad is the AMOUNT of unbound mAb in the depot
    d/dt(Ac) = (ka*Ad) + (Lp*L) - (C*CLp) - (C*L1*(1-sigma1)) - (C*L2*(1-sigma2)) ; # nmoles/seconds; unbound (free) AMOUNT of mAb in plasma
    d/dt(Att) = (C*L1*(1-sigma1)) - (Tt*L1*(1-sigmaL)) - (kon_tt * Att * Rtt) + (koff * ARtt) ; # nmoles/seconds; unbound (free) AMOUNT of mAb in tight tissue
    d/dt(Alk) = (C*L2*(1-sigma2)) - (Lk*L2*(1-sigmaL)) - (kon_lk * Alk * Rlk) + (koff * ARlk) ; # nmoles/seconds; unbound (free) AMOUNT of mAb in leaky tissue
    d/dt(Alym) = (Tt*L1*(1-sigmaL)) + (Lk*L2*(1-sigmaL)) - (Lp*L) ; # nmoles/seconds; unbound (free) AMOUNT of mAb in lymph
    d/dt(Rtt) = ksyn_tt - (kon_tt * Att * Rtt) + (koff * ARtt) - (kdeg * Rtt) ; # nmoles/seconds; unbound (free) AMOUNT of receptor in tight tissue
    d/dt(Rlk) = ksyn_lk - (kon_lk * Alk * Rlk) + (koff * ARlk) - (kdeg * Rlk) ; # nmoles/seconds; unbound (free) AMOUNT of receptor in leaky tissue
    d/dt(ARtt) = (kon_tt * Att * Rtt) - ((kint + koff) * ARtt) ; # nmoles/seconds; complex mAb+receptor in tight tissue 
    d/dt(ARlk) = (kon_lk * Alk * Rlk) - ((kint + koff) * ARlk) ; # nmoles/seconds; complex mAb+receptor in tight tissue 
    
    # receptor occpuancy in tight ('M') tissue and leaky tissue calculated here
    Rtot_tt = ARtt + Rtt
    Rtot_lk = ARlk + Rlk 
    ROtight = (ARtt) / (ARtt + Rtt); # % of Receptor occupancy in tight tissue 
    ROleaky = (ARlk) / (ARlk + Rlk); # % of Receptor occupancy in leaky tissue
    
    # final value of the concentration of drug in the plasma using the additive error model
    cp = log(C)
    
    # final value of the concentration of drug in the muscle (tight tissue) using the additive error model
    cm = log(Tt)`

# Additive model set up (observed data for cmpt 2 (central), cmpt 3 (tight tissue)
    model.add = modgen %>% model(cp ~ add(add.ct) | Ac, cm ~ add(add.cm) | Att, append = TRUE) %>%
    nlmixr2::ini(add.ct = 0.1, add.cm = 0.1)
    
    modellist = c("model.add")
    fitlist = list()

    for (i in 1:length(modellist)) {
                 fitlist[i] = list(nlmixr2(eval(parse(text = modellist[i])), 
                            dataFrame, 
                            est = "saem",
                            control = list(print = 50),
                            table = tableControl(cwres = T)))}

**Comment by mattfidler** [comment, marked as answer]
>  How do I elect a different solver from lsoda? 

You can choose`dop853` instead:

```r
...
control=list(print=50, rxControl(method="dop853"))
...
```

> Is it possible that my parameter estimates are so far off that the model does not converge?

It is possible, but ODE convergence has little to do with model convergence.  It is simply trying to search in a difficult to solve region.  `nlmixr2` tries to handle this as gracefully as it can, but it may not work.  You can try to adjust some of the solver settings yourself to see if that helps with `control=list(rxControl=rxControl(sigdig=12))` to add more accuracy to the solver for instance.  `hmax` and `hmin` can also be adjusted if you want to stick with `liblsoda`.
